// @google-cloud/functions/build/src/index.js
"use strict";
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by synthtool. **
// ** https://github.com/googleapis/synthtool **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.protos = exports.CloudFunctionsServiceClient = exports.v2beta = exports.v2alpha = exports.v2 = exports.v1 = void 0;
const v1 = require("./v1");
exports.v1 = v1;
const v2 = require("./v2");
exports.v2 = v2;
const v2alpha = require("./v2alpha");
exports.v2alpha = v2alpha;
const v2beta = require("./v2beta");
exports.v2beta = v2beta;
const CloudFunctionsServiceClient = v1.CloudFunctionsServiceClient;
exports.CloudFunctionsServiceClient = CloudFunctionsServiceClient;
exports.default = { v1, v2, v2alpha, v2beta, CloudFunctionsServiceClient };
const protos = require("../protos/protos");
exports.protos = protos;
